common/js/maps.ts

```diff
@@ -0,0 +1,37 @@
+// Utility functions for use on Maps
+
+/**
+ * Takes a map and appends a new item to an array at a given key. If the key doesn't exist, a new
+ * array with the item is created and inserted at the key.
+ *
+ * Mutates the input map instead of returning a new map.
+ */
+export function appendToMapArray<K, T>(map: Map<K, T[]>, key: K, item: T) {
```

Comment:
nit: to me, "MapArray" implies "an array of Maps". I think a better term for this is "MultiMap" (that's how C++ and Java refer to it)

Reply:
https://en.wikipedia.org/wiki/Multimap It has a wiki entry, I didn't make it up :P